### PR TITLE
Add module arguments to Module and it's subclasses

### PR DESCRIPTION
**Pull request by emilmelnikov**
Arguments are key-value pairs that are passed to `__init__`
and define the module behavior.
Saving this information allows creating new instances
of the same modules, similar to `tensor.new_tensor`.

All module implementations in PyTorch are changed to support
arguments, with the exception of `Container`, `Sequential`,
`ModuleList` and `ParameterList` in `torch/nn/modules/container.py`.

This PR implements feature request https://github.com/pytorch/pytorch/issues/7620.

### CODE REVIEW

**Comment by SsnL** [comment]
I wonder if we can do this using some metaclass trick rather than hardcoding everything.

**Comment by emilmelnikov** [comment]
Just obtaining all arguments from `__init__` is trivial with metaclasses:

```
class ModuleMeta(type):
    def __call__(self, *args, **kwargs):
        obj = super(ModuleMeta, self).__call__(*args, **kwargs)
        obj._arguments = (args, kwargs)
        return obj
```

However, how to handle `Sequential`, where arguments are modules themselves, and other non-trivial cases?

For modules that act as containers the ideal solution would be to somehow extract all arguments from the modules inside that container, and create a tree-like dict structure. I guess this even could be saved as JSON as a NN definition.

**Comment by apaszke** [comment]
+1 for a metaclass trick. it's a bit involved, because we'd need to retrieve the constructor arg names, but it would make the change pretty small

**Comment by SsnL** [comment]
Yeah. Getting the arg name can be a bit work. IIRC, we don't have any built-in modules that take in tensors, but if we do, we need to filter those args out too (do we?).

**Comment by apaszke** [comment]
Embedding has a private `_weights` argument that's used when you use `from_pretrained`

**Review by ezyang — CHANGES_REQUESTED** [review]
Use a metaclass instead

**Comment by emilmelnikov** [comment]
The metaclass `torch.nn._ModuleMeta` saves positional and keyword arguments to the instance's attributes `_args` and `_kwargs`. However, there are some issues:

1. We need to "normalize" incoming args (replace positional with keyword args whenever possible), because `new_module` accepts keyword args, and we need to merge old and new args. Currently, the quick-and-dirty implementation uses `inspect.getfullargspec` to do this. Unfortunately, the better solution with `inspect.Signature` is unavailable in Python 2.

2. `_ModuleMeta` conflicts with the jit-related `ScriptMeta` metaclass. The latter has been made a subclass of the former, but this change is a potentially breaking one.

3. It's not clear how to clone container modules (e.g. `Sequential`). For now, these modules have their own metaclass that disables parameter saving.

File: torch/nn/modules/module.py [hunk 1]
```diff
@@ -20,7 +23,24 @@ def _addindent(s_, numSpaces):
...
+    def __call__(cls, *args, **kwargs):
+        # Move positional to keyword arguments, but check for name clashes first.
+        # If there are any, call __init__ as-is to get a standard exception.
+        argnames = inspect.getargspec(cls.__init__).args[1:]
+        for name, value in zip(argnames, args):
+            kwargs.setdefault(name, value)
+        args = args[len(argnames):]
+        obj = type.__call__(cls, *args, **kwargs)
+        obj.__dict__.update(_args=args, _kwargs=kwargs)
+        return obj
```
Comment (apaszke) [inline torch/nn/modules/module.py]: Can you make it work only with `torch.*` modules? I'm a bit afraid of supporting it in user-defined modules, since constructors can take in some pretty heavyweight args like pretrained weights, etc.

File: torch/nn/modules/module.py [hunk 2]
```diff
@@ -20,7 +23,24 @@ def _addindent(s_, numSpaces):
...
+    """
+    def __call__(cls, *args, **kwargs):
+        # Move positional to keyword arguments, but check for name clashes first.
+        # If there are any, call __init__ as-is to get a standard exception.
+        argnames = inspect.getargspec(cls.__init__).args[1:]
+        for name, value in zip(argnames, args):
+            kwargs.setdefault(name, value)
+        args = args[len(argnames):]
+        obj = type.__call__(cls, *args, **kwargs)
+        obj.__dict__.update(_args=args, _kwargs=kwargs)
```
Comment (apaszke) [inline torch/nn/modules/module.py]: Can you make sure `_args` and `_kwargs` are not already defined in the `__dict__`? It's a simple assert

File: torch/nn/modules/module.py [hunk 3]
```diff
@@ -150,6 +170,29 @@ def register_parameter(self, name, param):
...
 
+    def new_module(self, **kwargs):
+        r"""Create a new instance of this module.
+
+        The new instance will have different, freshly initialized
+        parameter values.
+        Module arguments will be the same, except the ones specified
+        in the positional and keyword arguments of this method.
+
+        Some modules, like Sequential, cannot be copied.
```
Comment (apaszke) [inline torch/nn/modules/module.py]: Well it's not really a copy, so calling it like that is a bit misleading.

File: torch/nn/modules/module.py [hunk 4]
```diff
@@ -150,6 +170,29 @@ def register_parameter(self, name, param):
...
+
+        Args:
+            **kwargs: keyword arguments to replace in the new module.
+
+        Returns:
+            new instance of this module.
```
Comment (apaszke) [inline torch/nn/modules/module.py]: Can you add an example to showcase when is that useful?

File: torch/jit/__init__.py [hunk 5]
```diff
@@ -527,7 +527,7 @@ def _get_valid_constant(v):
 # resolve references to `self.param` or `self.module`.
 
 
-class ScriptMeta(type(torch._C.ScriptModule)):
+class ScriptMeta(type(torch._C.ScriptModule), _ModuleMeta):
```
Comment (apaszke) [inline torch/jit/__init__.py]: I'm not convinced we want to add this one here. `ScriptModules` are generally for user-defined ones. See my argumentation below

**Comment by SsnL** [comment]
This also deserves some test in `test_nn`

**Comment by yf225** [comment]
@emilmelnikov Let us know if you have time to address the reviewers' comments, and then we can merge the PR. Thanks!
